src/referral_storage.cairo

```diff
@@ -0,0 +1,110 @@
+use starknet::{ContractAddress,ClassHash};
+use zeroable::Zeroable;
+
+
+#[starknet::interface]
+trait IReferralStorage<TContractState> {
+    fn get_trader_referral_code(
+         self: @TContractState,
+        _account: ContractAddress,
+    ) -> ContractAddress;
+
+    fn set_trader_referral_code(
+        ref self: TContractState,
+        _code: ContractAddress,
+    );
+
+    fn upgrade(ref self: TContractState, new_class_hash: ClassHash);
+}
+
+
+#[starknet::contract]
+mod ReferralStorage {
+    use starknet::ContractAddress;
+    use starknet::get_caller_address;
+    use zeroable::Zeroable;
+    use starknet::ClassHash;
+    use openzeppelin::access::ownable::OwnableComponent;
+    use openzeppelin::upgrades::UpgradeableComponent;
+    use openzeppelin::upgrades::interface::IUpgradeable;
+
+
+    component!(path: UpgradeableComponent, storage: upgradeable_storage, event: UpgradeableEvent);
+    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);
+
+    #[abi(embed_v0)]
+    /// Ownable
+    impl OwnableImpl = OwnableComponent::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+
+    /// Upgradeable
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        SetTraderReferralCode: SetTraderReferralCode,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event
+    }
+
+
+    #[derive(Drop, starknet::Event)]
+    struct SetTraderReferralCode {
+        account: ContractAddress,
+        code: ContractAddress,
+    }
+
+    #[storage]
+    struct Storage {
+        // @notice Maps the trader to the referee code
+        trader_referral_codes: LegacyMap::<ContractAddress, ContractAddress>,
+        
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable_storage: UpgradeableComponent::Storage,
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, owner: ContractAddress) {
+        self.ownable.initializer(owner);
+    }
+
+    #[abi(embed_v0)]
+    impl ReferralStorage of super::IReferralStorage<ContractState> {
+
+        fn get_trader_referral_code(
+             self: @ContractState,
+            _account: ContractAddress,
+        ) -> ContractAddress {
+             self.trader_referral_codes.read(_account)
+        }
+
+        // @notice Sets the code to be referred by the trader
+        // @param _code The code to set
+        // @dev Trader can set the code only once
+        // @dev Code owner cannot set the code for himself
+        fn set_trader_referral_code(
```

Comment:
can rename to something like 'add_referrer'.

Reply:
changed it to set_referrer